File: tools/mdt_decoder.lua
#!/usr/bin/env lua5.1
--[[
MDT Route String Decoder (Standalone Lua)
==========================================

This script decodes Mythic Dungeon Tools route strings outside of World of Warcraft.
It reimplements the necessary parts of LibDeflate and AceSerializer.

Requirements:
- Lua 5.1 (same version as WoW uses)
- lua-zlib (luarocks install lua-zlib)

Usage:
lua mdt_decoder.lua route_string.txt
lua mdt_decoder.lua "!fBvtpUjmq0FrbH)aS9X..."

Note: This is a simplified implementation for educational purposes.
For production use, consider the Python/Node.js/C# versions.
--]]

local zlib = require('zlib')
local json = require('json') or require('dkjson') or require('cjson')

-- If no JSON library is available, create a simple one
if not json then
  json = {
    encode = function(obj)
      if type(obj) == "table" then
        local parts = {}
        local is_array = true
        local max_index = 0

        for k, v in pairs(obj) do
          if type(k) ~= "number" then
            is_array = false
            break
          else
            max_index = math.max(max_index, k)
          end
        end

        if is_array then
          local array_parts = {}
          for i = 1, max_index do
            array_parts[i] = json.encode(obj[i] or "null")
          end
          return "["..table.concat(array_parts, ",").."]"
        else
          for k, v in pairs(obj) do
            table.insert(parts, '"'..tostring(k)..'":'..json.encode(v))
          end
          return "{"..table.concat(parts, ",").."}"
        end
      elseif type(obj) == "string" then
        return '"'..obj:gsub('"', '\\"')..'"'
      elseif type(obj) == "number" then
        return tostring(obj)
      elseif type(obj) == "boolean" then
        return obj and "true" or "false"
      else
        return '"'..tostring(obj)..'"'
      end
    end
  }
end

-- Custom Base64 alphabet used by MDT
local B64_ALPHABET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789()"

-- Create decode table
local B64_DECODE = {}
for i = 1, #B64_ALPHABET do
  B64_DECODE[B64_ALPHABET:sub(i, i)] = i - 1
end

-- Bit manipulation functions (for Lua 5.1 compatibility)
local function bit_and(a, b)
  local result = 0
  local bitval = 1
  while a > 0 and b > 0 do
    if a % 2 == 1 and b % 2 == 1 then
      result = result + bitval
    end
    bitval = bitval * 2
    a = math.floor(a / 2)
    b = math.floor(b / 2)
  end
  return result
end

local function bit_lshift(a, b)
  return a * (2 ^ b)
end

local function bit_rshift(a, b)
  return math.floor(a / (2 ^ b))
end

-- Decode custom Base64 string to binary data
local function decode_custom_b64(encoded_str)
  local decoded = {}
  local bitfield = 0
  local bitfield_len = 0
  local decoded_size = 0

  for i = 1, #encoded_str do
    local ch = B64_DECODE[encoded_str:sub(i, i)]
    if ch then
      bitfield = bitfield + bit_lshift(ch, bitfield_len)
      bitfield_len = bitfield_len + 6

      while bitfield_len >= 8 do
        decoded_size = decoded_size + 1
        decoded[decoded_size] = string.char(bit_and(bitfield, 255))
        bitfield = bit_rshift(bitfield, 8)
        bitfield_len = bitfield_len - 8
      end
    end
  end

  return table.concat(decoded)
end

-- Decompress deflate data
local function decompress_data(compressed_data)
  local success, result = pcall(function()
    local stream = zlib.inflate()
    local decompressed, eof, bytes_in, bytes_out = stream(compressed_data)

    if decompressed then
      return decompressed
    else
      error("Decompression failed")
    end
  end)

  if success then
    return result
  else
    error("Decompression failed: "..tostring(result))
  end
end

-- Parse AceSerializer format data (simplified)
local function parse_ace_serialized_data(data)
  local result = {
    format = "AceSerializer",
    raw_data_hex = "",
    parsed_data = "Partial parsing - full AceSerializer implementation needed"
  }

  -- Convert to hex for inspection
  for i = 1, #data do
    result.raw_data_hex = result.raw_data_hex..string.format("%02x", string.byte(data, i))
  end

  -- Try to extract some basic information
  if data:find("week") then
    result.contains_week_info = true
  end
  if data:find("objects") then
    result.contains_objects = true
  end
  if data:find("pulls") then
    result.contains_pulls = true
  end

  -- Try to find patterns
  local dungeon_pattern = data:match("dungeonIdx.-(%d+)")
  if dungeon_pattern then
    result.dungeon_id = tonumber(dungeon_pattern)
  end

  local week_pattern = data:match("week.-(%d+)")
  if week_pattern then
    result.week = tonumber(week_pattern)
  end

  return result
end

-- Decode an MDT route string
local function decode_mdt_route_string(route_string)
  local success, result = pcall(function()
    -- Step 1: Remove the leading "!" marker
    local encoded_data = route_string
    if route_string:sub(1, 1) == "!" then
      encoded_data = route_string:sub(2)
    end

    -- Step 2: Decode custom Base64
    print("Decoding Base64...")
    local compressed_data = decode_custom_b64(encoded_data)
    print("Decoded "..#compressed_data.." bytes")

    -- Step 3: Decompress deflate data
    print("Decompressing data...")
    local decompressed_data = decompress_data(compressed_data)
    print("Decompressed to "..#decompressed_data.." bytes")

    -- Step 4: Parse serialized data
    print("Parsing serialized data...")
    local parsed_data = parse_ace_serialized_data(decompressed_data)

    -- Step 5: Create result structure
    return {
      metadata = {
        original_length = #route_string,
        compressed_length = #compressed_data,
        decompressed_length = #decompressed_data,
        compression_ratio = math.floor((#compressed_data / #decompressed_data) * 100) / 100
      },
      route_data = parsed_data
    }
  end)

  if success then
    return result
  else
    return {
      error = "Failed to decode route string: "..tostring(result),
      original_string = #route_string > 100 and (route_string:sub(1, 100).."...") or route_string
    }
  end
end

-- Main function
local function main(args)
  args = args or arg or {}

  if #args ~= 1 then
    print("Usage: lua mdt_decoder.lua <route_string_or_file>")
    print("Example: lua mdt_decoder.lua rt1.txt")
    print("Example: lua mdt_decoder.lua '!fBvtpUjmq0FrbH)aS9X...'")
    return
  end

  local input_arg = args[1]
  local route_string

  -- Check if it's a file or direct string
  if input_arg:match("%.txt$") or input_arg:match("[/\\]") then
    -- It's a file path
    local file = io.open(input_arg, "r")
    if not file then
      print("Error: File '"..input_arg.."' not found")
      return
    end

    route_string = file:read("*all"):gsub("%s+$", "")
    file:close()
  else
    -- It's a direct string
    route_string = input_arg
  end

  -- Decode the route string
  local result = decode_mdt_route_string(route_string)

  -- Output as JSON
  print(json.encode(result))
end

-- Run main function
main(arg)

File: tools/mdt_decoder.py
#!/usr/bin/env python3
"""
MDT Route String Decoder
========================

This script decodes Mythic Dungeon Tools route strings and converts them to JSON.
The route strings are encoded using:
1. Custom Base64 encoding (a-z, A-Z, 0-9, (, ))
2. Deflate compression
3. Custom serialization format

Requirements:
- pip install zlib (usually built-in)
- No external dependencies needed

Usage:
python mdt_decoder.py route_string.txt
python mdt_decoder.py "!fBvtpUjmq0FrbH)aS9X..."
"""

import sys
import zlib
import json
import re
from typing import Dict, Any, List, Union

# Custom Base64 alphabet used by MDT
B64_ALPHABET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789()"

def create_b64_decode_table():
    """Create lookup table for Base64 decoding"""
    return {char: idx for idx, char in enumerate(B64_ALPHABET)}

def decode_custom_b64(encoded_str: str) -> bytes:
    """Decode custom Base64 string to bytes"""
    decode_table = create_b64_decode_table()

    # Remove padding and prepare for decoding
    padding_removed = encoded_str.rstrip('=')

    # Convert to binary string
    binary_str = ''
    for char in padding_removed:
        if char in decode_table:
            binary_str += format(decode_table[char], '06b')

    # Convert binary to bytes
    byte_data = bytearray()
    for i in range(0, len(binary_str) - 7, 8):
        byte_chunk = binary_str[i:i+8]
        if len(byte_chunk) == 8:
            byte_data.append(int(byte_chunk, 2))

    return bytes(byte_data)

def decompress_data(compressed_data: bytes) -> bytes:
    """Decompress deflate-compressed data"""
    try:
        return zlib.decompress(compressed_data)
    except zlib.error as e:
        raise ValueError(f"Decompression failed: {e}")

def parse_ace_serialized_data(data: str) -> Dict[str, Any]:
    """
    Parse AceSerializer format data
    This is a simplified parser - the actual format is more complex
    """
    try:
        # Remove the AceSerializer header/version info
        # AceSerializer format typically starts with version info
        if data.startswith('\x01\x00'):
            data = data[2:]  # Remove version bytes

        # This is a simplified parser - for full compatibility,
        # you'd need to implement the full AceSerializer protocol
        # For now, we'll extract what we can

        result = {
            "format": "AceSerializer",
            "raw_data": data.encode('unicode_escape').decode('ascii'),
            "parsed_data": "Partial parsing - full AceSerializer implementation needed"
        }

        # Try to extract some basic information
        if "week" in data:
            result["contains_week_info"] = True
        if "objects" in data:
            result["contains_objects"] = True
        if "pulls" in data:
            result["contains_pulls"] = True

        return result

    except Exception as e:
        return {
            "error": f"Failed to parse serialized data: {e}",
            "raw_data": data.encode('unicode_escape').decode('ascii')
        }

def decode_mdt_route_string(route_string: str) -> Dict[str, Any]:
    """
    Decode an MDT route string to JSON-compatible format
    """
    try:
        # Step 1: Remove the leading "!" marker
        if route_string.startswith('!'):
            encoded_data = route_string[1:]
        else:
            encoded_data = route_string

        # Step 2: Decode custom Base64
        print("Decoding Base64...")
        compressed_data = decode_custom_b64(encoded_data)
        print(f"Decoded {len(compressed_data)} bytes")

        # Step 3: Decompress deflate data
        print("Decompressing data...")
        decompressed_data = decompress_data(compressed_data)
        print(f"Decompressed to {len(decompressed_data)} bytes")

        # Step 4: Parse serialized data
        print("Parsing serialized data...")
        serialized_str = decompressed_data.decode('latin-1')  # Use latin-1 for binary data
        parsed_data = parse_ace_serialized_data(serialized_str)

        # Step 5: Create result structure
        result = {
            "metadata": {
                "original_length": len(route_string),
                "compressed_length": len(compressed_data),
                "decompressed_length": len(decompressed_data),
                "compression_ratio": round(len(compressed_data) / len(decompressed_data), 2)
            },
            "route_data": parsed_data
        }

        return result

    except Exception as e:
        return {
            "error": f"Failed to decode route string: {e}",
            "original_string": route_string[:100] + "..." if len(route_string) > 100 else route_string
        }

def main():
    if len(sys.argv) != 2:
        print("Usage: python mdt_decoder.py <route_string_or_file>")
        print("Example: python mdt_decoder.py rt1.txt")
        print("Example: python mdt_decoder.py '!fBvtpUjmq0FrbH)aS9X...'")
        return

    input_arg = sys.argv[1]

    # Check if it's a file or direct string
    if input_arg.endswith('.txt') or '/' in input_arg or '\\' in input_arg:
        # It's a file path
        try:
            with open(input_arg, 'r', encoding='utf-8') as f:
                route_string = f.read().strip()
        except FileNotFoundError:
            print(f"Error: File '{input_arg}' not found")
            return
        except Exception as e:
            print(f"Error reading file: {e}")
            return
    else:
        # It's a direct string
        route_string = input_arg

    # Decode the route string
    result = decode_mdt_route_string(route_string)

    # Output as JSON
    print(json.dumps(result, indent=2, ensure_ascii=False))

if __name__ == "__main__":
    main()
